d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -260,22 +269,64 @@ void handleD2URIMapResponse(DiscoveryResponseData data)
         {
           _log.warn("Received a D2URIMap response with no data, resource is : {}", resourceName);
         }
-        updates.put(resourceName, new D2URIMapUpdate(resource.getVersion(), nodeData));
+        updates.put(resourceName, new D2URIMapUpdate(nodeData));
       }
       catch (InvalidProtocolBufferException e)
       {
         _log.warn("Failed to unpack D2URIMap response", e);
         errors.add("Failed to unpack D2URIMap response");
         // Assume that the resource doesn't exist if it cannot be deserialized instead of simply ignoring it. This way
         // any call waiting on the response can be satisfied instead of timing out.
-        updates.put(resourceName, new D2URIMapUpdate(resource.getVersion(), null));
+        updates.put(resourceName, new D2URIMapUpdate(null));
       }
     }
     sendAckOrNack(data.getResourceType(), data.getNonce(), errors);
     handleResourceUpdate(updates, data.getResourceType());
     handleResourceRemoval(data.getRemovedResources(), data.getResourceType());
   }
 
+  @VisibleForTesting
+  void handleD2URIResponse(DiscoveryResponseData data)
+  {
+    Map<String, D2URICollectionUpdate> updates = new HashMap<>();
+    List<String> errors = new ArrayList<>();
+
+    for (Resource resource : data.getResourcesList())
+    {
+      try
+      {
+        XdsD2.D2URI uri = resource.getResource().unpack(XdsD2.D2URI.class);
+        getCollectionUpdate(updates, resource.getName())
+            .addUri(uriFromResourceUrn(resource.getName()), uri);
+      }
+      catch (InvalidProtocolBufferException e)
+      {
+        _log.warn("Failed to unpack D2URI glob collection response", e);
+        errors.add("Failed to unpack D2URI glob collection response");
+      }
+    }
+    for (String removedResource : data.getRemovedResources())
+    {
+      getCollectionUpdate(updates, removedResource)
+          .addRemovedUri(removedResource);
+    }
+    sendAckOrNack(data.getResourceType(), data.getNonce(), errors);
```

Comment:
we should log about empty D2URICollectionUpdate (similar to line 237 and 270).

```suggestion
    for (Map.Entry<String, D2URICollectionUpdate > entry : updates.entrySet())
    {
       if (isEmptyCollectionUpdate(entry.value()))
       {
          _log.warn("Received D2URICollectionUpdate with no data for resource: {}", entry.key());
       }
    }
    sendAckOrNack(data.getResourceType(), data.getNonce(), errors);
```

isEmptyCollectionUpdate could be a helper function (to be used at line 527 as well).

Reply:
synced offline. Actually empty maps should be valid for cluster uris maps since the uris under a cluster could have all de-announced legitimately. (as opposed to removing the UriMap resource entirely should be ignored).